src/Microsoft.AspNetCore.Server.Kestrel.Https/HttpsConnectionAdapter.cs

```diff
@@ -37,6 +44,11 @@ public HttpsConnectionAdapter(HttpsConnectionAdapterOptions options, ILoggerFact
                 throw new ArgumentException(HttpsStrings.ServiceCertificateRequired, nameof(options));
             }
 
+            // capture the certificate now so it can be switched after validation
+            _serverCertificate = options.ServerCertificate;
+
+            ValidateEku(_serverCertificate);
```

Comment:
I'm not a fan of having logic in a ctor. Have you considered validating in the `HttpConnectionAdapterOptions.ServerCertificate` setter?